nixos/static-web-server: create module which uses upstream systemd units

mac-chaffee opened the pull request:
###### Description of changes

This commit creates a nixos module for [static-web-server](https://static-web-server.net/). The module uses upstream systemd units to start static-web-server. It also includes options for configuring static-web-server. 

I also added a test which will install static-web-server with a custom port number, custom root dir, and custom config file, then curl it to see if it responds with a 200.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

mac-chaffee commented:
@figsoda friendly ping, freshly rebased :)

figsoda commented on nixos/modules/services/web-servers/static-web-server.nix:
```diff
@@ -0,0 +1,70 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
would be nice to remove this

figsoda commented on pkgs/servers/static-web-server/default.nix:
```diff
@@ -24,6 +24,14 @@ rustPlatform.buildRustPackage rec {
 
+  # Need to copy in the systemd units for systemd.packages to discover them
+  postInstall = ''
+    install -Dm444 -t $out/lib/systemd/system/ systemd/static-web-server.service
+    install -Dm444 -t $out/lib/systemd/system/ systemd/static-web-server.socket
```
```suggestion
    install -Dm444 -t $out/lib/systemd/system/ systemd/static-web-server.{service,socket}
```

figsoda commented on nixos/tests/web-servers/static-web-server.nix:
```diff
@@ -0,0 +1,27 @@
+import ../make-test-python.nix ({ pkgs, lib, ...} : {
+  name = "static-web-server";
+  meta = with pkgs.lib.maintainers; {
+    maintainers = [ mac-chaffee ];
```
```suggestion
  meta = {
    maintainers = with lib.maintainers; [ mac-chaffee ];
```

figsoda commented on nixos/tests/web-servers/static-web-server.nix:
```diff
@@ -0,0 +1,27 @@
+import ../make-test-python.nix ({ pkgs, lib, ...} : {
```
```suggestion
import ../make-test-python.nix ({ pkgs, lib, ... } : {
```

figsoda commented on maintainers/maintainer-list.nix:
```diff
@@ -9769,6 +9769,11 @@
       fingerprint = "1147 43F1 E707 6F3E 6F4B  2C96 B9A8 B592 F126 F8E8";
     }];
   };
+  mac-chaffee = {
```
please put this in a separate commit that precedes the other commit(s)
```
maintainers: add mac-chaffee
```

figsoda reviewed (commented):
sorry about the delay, lgtm overall, just some nitpicks

figsoda commented on nixos/tests/web-servers/static-web-server.nix:
```diff
@@ -0,0 +1,27 @@
+    };
+  };
+
+  testScript = ''
```
would be nice if this also tests the files served, perhaps if we change `root` to a store path

mac-chaffee committed:
maintainers: add mac-chaffee

mac-chaffee committed:
nixos/static-web-server: create module which uses upstream systemd units

This commit creates a nixos module for static-web-server.
The module uses upstream systemd units to start static-web-server.
It also includes options for configuring static-web-server.

figsoda commented:
@ofborg build static-web-server static-web-server.tests

figsoda reviewed (approved):
Thank you!

mac-chaffee commented:
Thanks for the review!

I also created a wiki page for using this module: https://nixos.wiki/wiki/Static_Web_Server

joseluisq commented:
JFYI  The package and the module is now mentioned in the SWS docs. Thanks!
https://github.com/static-web-server/static-web-server/commit/eb2887f8dfa0d850d274eed8087eaa07722acebb
